src/core/mac/sub_mac.cpp

```diff
@@ -1029,6 +1029,142 @@ void SubMac::StartTimerAt(Time aStartTime, uint32_t aDelayUs)
 #endif
 }
 
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE || OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+void SubMac::RadioSample(void)
+{
+#if OPENTHREAD_CONFIG_MAC_FILTER_ENABLE
+    VerifyOrExit(!mRadioFilterEnabled, IgnoreError(Get<Radio>().Sleep()));
+#endif
+
+    SetState(kStateRadioSample);
+
+    if (RadioSupportsReceiveTiming() || !RequestSample())
+    {
+#if !OPENTHREAD_CONFIG_MAC_CSL_DEBUG_ENABLE
+        IgnoreError(Get<Radio>().Sleep()); // Don't actually sleep for debugging
+#endif
+    }
+
+#if OPENTHREAD_CONFIG_MAC_FILTER_ENABLE
+exit:
+#endif
+    return;
+}
+
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE && OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+//----------------------------------------------------------------------------------------------------------------------
+// CSL receiver and WED listener case
+
+bool SubMac::IsRadioSampleEnabled(void) const { return IsCslEnabled() || mIsWedEnabled; }
+
+/*
+ * The radio state (receive/sleep) is determined by the request from both CSL and WED:
+ * <1> If both CSL and WED request to enter sleep state, the radio is set to sleep state.
+ * <2> If either CSL or WED requests to enter the receive state and the other requests to enter sleep state, the radio
+ *     is set to receive state using the channel that is requested to enter the receive state.
+ * <3> If both CSL and WED request to enter the receive state, the radio is set to the receive state using the CSL
+ *     channel.
+ *
+ * The diagram below illustrates how to set the radio state based on the request of WED and CSL.
+ *
+ * CSL   ------========------------========------------========------------========---
+ *             ^       ^
+ *             |       |
+ *             | RequestSleep()
+ *     RequestReceive()
+ *
+ * WED   -----------++++++++----------------++++++++----------------++++++++----------
+ *                  ^       ^
+ *                  |       |
+ *                  | RequestSleep()
+ *         RequestReceive()
+ *
+ * Radio ------========+++++-------========-++++++++---========-----+++++++========---
+ *             ^       ^    ^
+ *             |       |    |
+ *             |       | Radio::Sleep()
+ *             |  Radio::Receive(WedCh)
+ *      Radio::Receive(CslCh)
+ */
+void SubMac::RequestSleep(void)
+{
+    if (!mIsCslSampling && !mIsWedSampling)
```

Comment:
An alternative way can be to first calling `RequestSample()` and checking if it is starts rx/sampling (from its returned bool value), and if not then putting radio to sleep? What do you think? I feel make it simpler?

```c++
void SubMac::RequestSleep(void)
{
     if (!RequestSample())
     {
           IgnoreError(Get<Radio>().Sleep());
     }
}